File: src/main/java/org/gd/leetcode/p0450/Solution.py
# Task:
# https://leetcode.com/problems/delete-node-in-a-bst/
# 
# Iterative Solution:
# https://leetcode.com/problems/delete-node-in-a-bst/discuss/261868/
# 

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def find_min(slef, root: TreeNode) -> TreeNode:
        while root.left:
            root = root.left
        return root

    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        if root:
            if key < root.val:
                root.left = self.deleteNode(root.left, key)
            elif key > root.val:
                root.right = self.deleteNode(root.right, key)
            else:
                if not root.left:
                    return root.right

                if not root.right:
                    return root.left

                min_node: TreeNode = self.find_min(root.right)
                root.val = min_node.val
                root.right = self.deleteNode(root.right, root.val)

        return root

File: src/main/java/org/gd/leetcode/p0450/Solution.java
package org.gd.leetcode.p0450;

import org.gd.common.Repeat;
import org.gd.leetcode.common.LeetCode;
import org.gd.leetcode.common.TreeNode;

/**
 * https://leetcode.com/problems/delete-node-in-a-bst/
 */
@Repeat("delete node from BST")
@LeetCode(difficulty = LeetCode.Level.MEDIUM, tags = { LeetCode.Tags.TREE, LeetCode.Tags.BINARY_SEARCH_TREE })
class Solution {

    private static TreeNode findMin(TreeNode node) {
        while (node.left != null)
            node = node.left;
        return node;
    }

    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null)
            return null;

        if (key < root.val) {
            root.left = deleteNode(root.left, key);
        } else if (key > root.val) {
            root.right = deleteNode(root.right, key);
        } else {

            if (root.left == null)
                return root.right;

            if (root.right == null)
                return root.left;

            TreeNode min = findMin(root.right);
            root.val = min.val;
            root.right = deleteNode(root.right, root.val);
        }

        return root;
    }
}
